### ISSUE TITLE
[Firebase/Performance] `GULObjectSwizzler` crashes on `objc_disposeClassPair`.

### ISSUE THREAD

**Issue by asafkorem** [firebase/firebase-ios-sdk#9083]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: `13.1`
  * Firebase SDK version: `8.10`
  * Installation method: Any
  * Firebase Component: Firebase
  * Target platform(s): `iOS`

#### Steps to reproduce:

When using Firebase Performance along with [Detox Sync](https://github.com/wix/DetoxSync) (a framework used by [Detox](https://github.com/wix/Detox)), there is a double ISA Swizzling by the components. This causes a crash when `NSZombiesEnabled=NO`.

* Firebase Performance does ISA Swizzling (using [GULObjectSwizzler](https://github.com/google/GoogleUtilities/blob/main/GoogleUtilities/ISASwizzler/GULObjectSwizzler.m)).
* Detox Sync does ISA Swizzling, with a dynamic subclass of the generated class, using [DTXSwizzlingHelper](https://github.com/wix/DTXObjectiveCHelpers/blob/1f1b380775f604c61ec4eb8584df40f1d8487291/DTXSwizzlingHelper.h#L99).
* `GULObjectSwizzler` calls to dispose its own generated subclasses using `objc_disposeClassPair()`.
* Crash with the error message: `
objc_disposeClassPair: class 'fir_<some uuid>_<class name>' still has subclasses, including 'fir_<some uuid>_<class name>(__detox_sync_<class name>)'!
`

----

This issue seems to be related: https://github.com/google/GoogleUtilities/issues/15.
Also, another explanation of a similar issue, with zombie objects: https://github.com/firebase/firebase-ios-sdk/issues/8321#issuecomment-900321941 (instead Zombie objects, we have DetoxSync objects).

----

This branch of DetoxSync reproduces the issue: https://github.com/asafkorem/DetoxSync/tree/feature/reproduce-detox-sync-crash: build and run [ExampleApp/TimersTest.xcodeproj](https://github.com/asafkorem/DetoxSync/tree/feature/reproduce-detox-sync-crash/ExampleApp/TimersTest.xcodeproj) with `NSZombiesEnabled=NO`.

**Comment by mikehardy** [comment]
@paulb777 just to note this is of intense interest in react-native-firebase repo, our test harness is based entirely on Detox, and last time I did a survey of available e2e test tools, Detox still won, even with the horrific ios e2e crash frequency we're taking now. Would :heart: to see some sort of fix or remediation or something here :-)

**Comment by d4vidi** [comment]
Thanks @mikehardy, and sorry about the inconvenience around this thus far. Looks like we're really getting somewhere now.

**Comment by paulb777** [comment]
@asafkorem @mikehardy Thanks for the detailed report and background. We'll take a look.  It's one of our priorities to overhaul how Firebase uses swizzling. However, a resolution may still be months out.

In the meantime, is omitting FirebasePerformance when Detox is used a viable workaround?

**Comment by mikehardy** [comment]
It is possible but is equivalent to saying omit performance from e2e testing for us

**Comment by asafkorem** [comment]
Thanks for the response @paulb777, as @mikehardy mentioned it is possible but we'd rather avoid that, I think it would be better as a workaround to set some compiler flag that will only omit the [deallocation](https://github.com/google/GoogleUtilities/blob/616fac2626b6b2d1424d79a6f786b4e2ed1cfb49/GoogleUtilities/ISASwizzler/GULObjectSwizzler.m#L162) part that triggers this crash. This flag can be used when building the app for Detox tests.

**Comment by mikehardy** [comment]
@asafkorem is there any reason we can't ride-along with `NSZombiesEnabled=1` behavior - that is, make GULObjectSwizzler think we are running the zombies instrument, and thus avoid de-allocation.

Stated differently, perhaps the "NSZombiesEnabled" check could more properly be thought of as a "don't dispose generated classes" check, just with a too-specific name, but it would serve us as well :thinking: . Have you tried it?

**Comment by asafkorem** [comment]
@mikehardy yes, I thought about it too and already tried this approach. 
I wasn't able to set this environment variable using `xcodebuild` command :/

**Comment by asafkorem** [comment]
Oh, this makes more sense now.. thanks @mikehardy. I'll try this soon

**Comment by ncooke3** [comment]
cc: @visumickey

**Comment by mikehardy** [comment]
It appears to be working for me, if not completely, then "well". The first run I did I took a SIGSEGV in DetoxSync again, but then I was 10/10 on further runs (an incredible result, I was at maybe 1-2/10 success prior). I am using the special method of passing env vars through simctl as mentioned above. I'm about to make a commit to react-native-firebase that will have a github cross link so others may examine in case it's useful. Thanks for this trick, at least preliminary results show it appears workable to me

**Comment by shamilovtim** [comment]
I'm a little concerned about the SIGSEGV. Was that just a freak accident? At least with the DetoxSync crash there was some transparency into what it was. With SIGSEGVs I dislike that they're usually a disaster to debug

**Comment by mikehardy** [comment]
I'll need to wait to collect quite a few more results I think, before I want to spend anyone's time in this repo, possibly chasing ghosts.

**Comment by mikehardy** [comment]
Hmm, sadly I think statistics is a thing and I simply had an improbably long run of good results. With quite a few more runs throughout the day / evening, I still definitely have a problem with Detox-current on firebase-ios-sdk-current on iOS-simulator-current on both local macOS x86-64 VMs and up on github actions macOS runners :shrug: - really seemed like an improvement, and I have a hunch statistically it still might be, but there's no silver bullet with the NSZombiesEnabled=1, at least the way I'm setting it https://github.com/invertase/react-native-firebase/runs/4511769043?check_suite_focus=true :shrug:

**Comment by shamilovtim** [comment]
Yeah can confirm still seeing Signal 11 with `SIMCTL_CHILD_NSZombiesEnabled=1`. Is there an easy way of confirming that this property is being injected into the simulator build?

**Comment by asafkorem** [comment]
@mikehardy Adding `SIMCTL_CHILD_NSZombiesEnabled=1` before the `xcrun simctl launch` command doesn't seems to solve this (`NSZombiesEnabled` is still false).

**Comment by mikehardy** [comment]
Really sorry I took everyone down the wrong path with my plural-zombies typo! Thanks @asafkorem for seeing that and clearing it up. Excited to have this mitigated, my CI is running now in react-native-firebase based off your PR there Asaf

**Comment by shamilovtim** [comment]
@asafkorem  Nice catch!

**Comment by asafkorem** [comment]
@paulb777 @ncooke3 @visumickey 
Although passing the `NSZombieEnabled` flag does solves this problem as a workaround, I would have preferred to have a dedicated flag for disabling this class-disposing from firebase-performance and not to ride on the zombie-enabled flag, which might cause other side-effects like ignoring real memory issues on detox tests or unnecessary memory consumption, especially if a clean solution may only arrive in months. 🙏🏼

**Comment by asafkorem** [comment]
> .. I would have preferred to have a dedicated flag for disabling this class-disposing from firebase-performance and not to ride on the zombie-enabled flag, which might cause other side-effects ..

I've opened a PR for this: https://github.com/google/GoogleUtilities/pull/66

**Comment by paulb777** [comment]
Thanks @asafkorem!  GoogleUtilities 7.7.0 is now published to CocoaPods and SPM with your PR that disables class disposing when the `GULGeneratedClassDisposeDisabled` environment variable is set.

**Comment by asafkorem** [comment]
Thanks @paulb777! 
We'll replace the zombie flag with this one on Detox tests execution command. 
FYI @mikehardy.

**Comment by mikehardy** [comment]
Sample size of 10 runs which - for this bug, in the react-native-firebase harness - is strongly statistically significant already. Looks great, thanks for the ping @asafkorem

**Comment by asafkorem** [comment]
Although enabling this flag is a bit workaround, it's a pretty decent one. In my opinion, we can definitely live with that, considering the fact that this crash happens when running the apps on Detox tests environment, and this kind of memory consumption (not disposing the generated classes) should not particularly worry us in this case.

I opened this issue in the context of this bug, and I think we can close it now since it was resolved. Thanks guys!
